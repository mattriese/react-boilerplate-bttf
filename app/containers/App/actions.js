/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  GET_QUOTES,
  ADD_QUOTE,
  LOAD_QUOTES_SUCCESS,
  LOAD_QUOTES_ERROR,
} from './constants';

/**
 * Add new quote to quote list Dispatched from component to watcher saga
 *
 * @returns {object} an object with a type of ADD_QUOTE
 */
export function addQuote() {
  return {
    type: ADD_QUOTE,
  };
}

/**
 * Dispatched by worker saga after api loads default quotes
 * @returns {object}      An action object with a type of ??? passing the quotes
 */
export function getQuotes() {
  return {
    type: GET_QUOTES,
  };
}

/**
 * Dispatched when the quotes are loaded by the request saga
 *
 * @param  {array} quotes The quotes data
 *
 * @return {object}      An action object with a type of LOAD_QUOTES_SUCCESS passing the quotes
 */
export function quotesLoaded(quotes) {
  return {
    type: LOAD_QUOTES_SUCCESS,
    quotes,
  };
}

/**
 * Dispatched when loading the repositories fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_QUOTES_ERROR passing the error
 */
export function quoteLoadingError(error) {
  return {
    type: LOAD_QUOTES_ERROR,
    error,
  };
}
