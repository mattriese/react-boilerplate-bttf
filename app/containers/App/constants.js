/*
 * AppConstants

 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

export const GET_QUOTES = 'boilerplate/App/GET_QUOTES';
export const ADD_QUOTE = 'boilerplate/App/ADD_QUOTE';
export const LOAD_QUOTES_SUCCESS = 'boilerplate/App/LOAD_QUOTES_SUCCESS';
export const LOAD_QUOTES_ERROR = 'boilerplate/App/LOAD_QUOTES_ERROR';
export const NEWQUOTE_SAVE_SUCCESS = 'boilerplate/App/NEWQUOTE_SAVE_SUCCESS';
export const RESET_SUCCESS = 'boilerplate/App/RESET_SUCCESS';
export const RESET_ERROR = 'boilerplate/App/RESET_ERROR';
