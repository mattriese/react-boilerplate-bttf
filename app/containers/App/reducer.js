/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';
import {
  LOAD_QUOTES_ERROR,
  LOAD_QUOTES_SUCCESS,
  NEWQUOTE_SAVE_SUCCESS,
  RESET_SUCCESS,
  RESET_ERROR,
} from './constants';

// The initial state of the App
export const initialState = {
  loading: false,
  error: false,
  quotes: [],
  success: false,
};

/* eslint-disable default-case, no-param-reassign */
const appReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case LOAD_QUOTES_ERROR:
        draft.error = action.error;
        draft.loading = false;
        draft.success = false;
        break;

      case LOAD_QUOTES_SUCCESS:
        draft.quotes = action.quotes;
        draft.loading = false;
        draft.success = false;
        break;

      case NEWQUOTE_SAVE_SUCCESS:
        draft.quotes = [action.savedQuote, ...state.quotes];
        draft.loading = false;
        draft.success = true;
        break;

      case RESET_SUCCESS:
        draft.quotes = draft.quotes;
        draft.loading = false;
        draft.success = false;
        break;

      case RESET_ERROR:
        draft.quotes = draft.quotes;
        draft.error = false;
        draft.loading = false;
        draft.success = false;
        break;
    }
  });

export default appReducer;
