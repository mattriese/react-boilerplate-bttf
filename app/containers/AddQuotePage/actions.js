/*
 * AddQuote Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import { CHANGE_NEWQUOTE, RESET_NEWQUOTE } from './constants';

/**
 * Changes the input field of the form
 *
 * @param  {string} newQuote The new text of the input field
 *
 * @return {object} An action object with a type of CHANGE_NEWQUOTE
 */
export function changeNewQuote(newQuote) {
  return {
    type: CHANGE_NEWQUOTE,
    newQuote,
  };
}

/**
 * Resets the input field of the form
 *
 * @return {object} An action object with a type of RESET_NEWQUOTE
 */
export function resetNewQuote() {
  return {
    type: RESET_NEWQUOTE,
  };
}
