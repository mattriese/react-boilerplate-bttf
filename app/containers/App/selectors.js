/**
 * The global state selectors
 */

import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectGlobal = state => state.global || initialState;

const selectRouter = state => state.router;

const makeSelectQuotes = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.quotes,
  );

const makeSelectSuccess = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.success,
  );

const makeSelectCurrentUser = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.currentUser,
  );

const makeSelectLoading = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.loading,
  );

const makeSelectError = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.error,
  );

const makeSelectLocation = () =>
  createSelector(
    selectRouter,
    routerState => routerState.location,
  );

export {
  selectGlobal,
  makeSelectSuccess,
  makeSelectLoading,
  makeSelectError,
  makeSelectLocation,
  makeSelectQuotes,
  makeSelectCurrentUser,
};
