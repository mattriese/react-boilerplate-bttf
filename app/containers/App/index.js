/**
 *
 * App
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 */

import React from 'react';
import { Helmet } from 'react-helmet';
import styled from 'styled-components';
import { Switch, Route } from 'react-router-dom';

import HomePage from 'containers/HomePage/Loadable';
import AddQuotePage from 'containers/AddQuotePage/Loadable';
import NotFoundPage from 'containers/NotFoundPage/Loadable';
import Header from 'components/Header';
import GlobalStyle from '../../global-styles';

const AppWrapper = styled.div`
  max-width: calc(768px + 16px * 2);
  margin: 0 auto;
  display: flex;
  min-height: 100%;
  padding: 0 16px;
  flex-direction: column;
`;

export default function App() {
  return (
    <AppWrapper>
      <Helmet
        titleTemplate="%s - Back to the Future Quotes"
        defaultTitle="Back to the Future Quotes"
      >
        <meta name="description" content="Back to the Future Quotes" />
      </Helmet>
      <Header />
      <Switch>
        <Route exact path="/" component={HomePage} />
        <Route path="/addQuote" component={AddQuotePage} />
        <Route path="" component={NotFoundPage} />
      </Switch>
      <GlobalStyle />
    </AppWrapper>
  );
}
