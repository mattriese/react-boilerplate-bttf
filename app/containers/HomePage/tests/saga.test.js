/**
 * Tests for HomePage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { GET_QUOTES } from 'containers/App/constants';
import { quotesLoaded, quoteLoadingError } from 'containers/App/actions';

import githubData, { requestGetQuotes } from '../saga';

const newQuote = 'mxstbr';

/* eslint-disable redux-saga/yield-effects */
describe('requestGetQuotes Saga', () => {
  let getReposGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getReposGenerator = requestGetQuotes();

    const selectDescriptor = getReposGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getReposGenerator.next(newQuote).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the reposLoaded action if it requests the data successfully', () => {
    const response = [
      {
        name: 'First repo',
      },
      {
        name: 'Second repo',
      },
    ];
    const putDescriptor = getReposGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(quotesLoaded(response, newQuote)));
  });

  it('should call the repoLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getReposGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(quoteLoadingError(response)));
  });
});

describe('githubDataSaga Saga', () => {
  const githubDataSaga = githubData();

  it('should start task to watch for LOAD_REPOS action', () => {
    const takeLatestDescriptor = githubDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(GET_QUOTES, requestGetQuotes),
    );
  });
});
