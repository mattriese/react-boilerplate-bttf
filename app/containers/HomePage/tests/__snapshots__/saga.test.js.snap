// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requestGetQuotes Saga should call the repoLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3001/quotes",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`requestGetQuotes Saga should call the repoLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "quotes": undefined,
      "type": "boilerplate/App/LOAD_QUOTES_SUCCESS",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`requestGetQuotes Saga should dispatch the reposLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3001/quotes",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`requestGetQuotes Saga should dispatch the reposLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "quotes": undefined,
      "type": "boilerplate/App/LOAD_QUOTES_SUCCESS",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;
